http/README.md

```diff
@@ -0,0 +1,159 @@
+
+When this document says *HTTP* it is generally referring to *HTTP semantics*.
+
+## Interoperability with existing HTTP systems
+
+A goal of this spec is to allow libp2p to be able to interoperate with existing HTTP servers and clients. Care is taken in this document to not introduce anything that would break interoperability with existing systems.
+
+## HTTP Transport
+
+Nodes MUST use HTTPS (i.e., they MUST NOT use plaintext HTTP). It is RECOMMENDED to use HTTP/2 and HTTP/3.
+
+Nodes signal support for their HTTP transport using the `/http` component in
+their multiaddr. E.g., `/dns4/example.com/tls/http`. See the [HTTP multiaddr
+component spec](https://github.com/libp2p/specs/pull/550) for more details.
+
+## Namespace
+
+libp2p does not squat the global namespace. libp2p application protocols can be discovered by the [well-known resource](https://www.rfc-editor.org/rfc/rfc8615) `.well-known/libp2p`. This allows server operators to dynamically change the URLs of the application protocols offered, and not hard-code any assumptions how a certain resource is meant to be interpreted.
+
+```json
+
+{
+    "protocols": {
+        "/kad/1.0.0": {"path": "/kademlia/"},
+        "/ipfs-http/1.0.0": {"path": "/"},
+    }
+}
+```
+
+The resource contains a mapping of application protocols to their respective URL. For example, this configuration file would tell a client
+
+1. That the Kademlia protocol is available at `/kademlia` and
+2. The [IPFS Trustless Gateway API](https://specs.ipfs.tech/http-gateways/trustless-gateway/) is mounted at `/`.
+
+It is valid to expose a service at `/`. It is RECOMMENDED that implementations facilitate the coexistence of different service endpoints by ensuring that more specific URLs are resolved before less specific ones. For example, when registering handlers, more specific paths like `/kademlia/foo` should take precedence over less specific handler, such as `/`.
+
+## Peer ID Authentication
+
+When using the HTTP Transport, peer id authentication is optional. You only pay for it if you need it. This benefits use cases that don’t need peer authentication (e.g., fetching content addressed data) or authenticate some other way (not tied to libp2p peer ids).
+
+Peer ID authentication in the HTTP Transport follows a similar to pattern to how
+libp2p adds Peer ID authentication in WebTransport and WebRTC. We run the
+standard libp2p Noise handshake, but using `IX` for client and server
+authentication or `NX` for just server authentication.
+
+Note: This is just one form of Peer ID authentication. Other forms may be added
+in the future (with a different `WWW-Authenticate` value) or be added to the
+application protocols themselves.
+
+### Authentication flow
+
+1. The client initiates a request that it knows must be authenticated OR the client responds to a `401` with the header `WWW-Authenticate: Libp2p-Noise-IX` (The server MAY also include `Libp2p-Token` as an authentication scheme).
```

Comment:
For all headers we use here, we should define them as structured fields (RFC 8941). This will allow us to use standardized header parsers (e.g. https://github.com/dunglas/httpsfv for Go) do the parsing for us, instead of writing a bespoke parser for every field we define.

There might be some objection to not pulling in multibase here, but I think that using an IETF-specified way to do header encoding outweighs the benefits of multibase.

Reply:
I'm not opposed. It looks fairly straightforward and it seems like there are libraries out for JS, Rust, and Go at least.

But, to be clear, we aren't doing bespoke parsing. We're using protobufs here, and we encode it into a multibase. This is just like what we do everywhere else we need to send structured data over the wire.

With the current definition, SFV might not get us much since we still need to reach for a protobuf parser to read the [NoiseHandshake](https://github.com/libp2p/specs/tree/master/noise#the-libp2p-handshake-payload) payload and we need Multihash to parse the peer id. So it could only replace the multibase encoding. Changing "b<base32>" (or "m<base64>")  to ":<base64>:".

We could consider switching to a different auth scheme, but I think that's a bigger change, and I wanted to avoid introducing a new auth scheme.